cirq-core/cirq/optimizers/two_qubit_to_sqrt_iswap.py

```diff
@@ -0,0 +1,340 @@
+            raise ValueError(
+                f'the given gate cannot be decomposed into exactly '
+                f'{required_sqrt_iswap_count} sqrt-iSWAP gates.'
+            )
+        return [
+            _decomp_0_matrices,
+            _decomp_1sqrt_iswap_matrices,
+            _decomp_2sqrt_iswap_matrices,
+            _decomp_3sqrt_iswap_matrices,
+        ][required_sqrt_iswap_count](kak, atol=atol)
+    if _in_0_region(kak.interaction_coefficients, weyl_tol=atol / 10):
+        return _decomp_0_matrices(kak, atol)
+    elif _in_1sqrt_iswap_region(kak.interaction_coefficients, weyl_tol=atol / 10):
+        return _decomp_1sqrt_iswap_matrices(kak, atol)
+    elif _in_2sqrt_iswap_region(kak.interaction_coefficients, weyl_tol=atol / 10):
+        return _decomp_2sqrt_iswap_matrices(kak, atol)
+    return _decomp_3sqrt_iswap_matrices(kak, atol)
+
+
+def _in_0_region(
+    interaction_coefficients: Tuple[float, float, float],
+    weyl_tol: float = 1e-8,
+) -> bool:
+    """Tests if (x, y, z) ~= (0, 0, 0) assuming x, y, z are canonical."""
+    x, y, z = interaction_coefficients
+    return abs(x) <= weyl_tol and abs(y) <= weyl_tol and abs(z) <= weyl_tol
+
+
+def _in_1sqrt_iswap_region(
+    interaction_coefficients: Tuple[float, float, float],
+    weyl_tol: float = 1e-8,
+) -> bool:
+    """Tests if (x, y, z) ~= (π/8, π/8, 0), assuming x, y, z are canonical."""
+    x, y, z = interaction_coefficients
+    return abs(x - np.pi / 8) <= weyl_tol and abs(y - np.pi / 8) <= weyl_tol and abs(z) <= weyl_tol
+
+
+def _in_2sqrt_iswap_region(
+    interaction_coefficients: Tuple[float, float, float],
+    weyl_tol: float = 1e-8,
+) -> bool:
+    """Tests if (x, y, z) is inside or within weyl_tol of the volume
+    x >= y + |z| assuming x, y, z are canonical.
+
+    References:
+        Towards ultra-high fidelity quantum operations: SQiSW gate as a native
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    x, y, z = interaction_coefficients
+    # Lemma 1 of the paper
+    return x + weyl_tol >= y + abs(z)
+
+
+def _in_3sqrt_iswap_region(
+    interaction_coefficients: Tuple[float, float, float],
+    weyl_tol: float = 1e-8,
+) -> bool:
+    """Any two-qubit operation is decomposable into three SQRT_ISWAP gates.
+
+    References:
+        Towards ultra-high fidelity quantum operations: SQiSW gate as a native
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    return True
+
+
+def _decomp_0_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 0-SQRT_ISWAP decomposition.
+
+    Assumes canonical x, y, z and (x, y, z) = (0, 0, 0) within tolerance.
+    """
+    return [
+        (
+            kak.single_qubit_operations_after[0] @ kak.single_qubit_operations_before[0],
+            kak.single_qubit_operations_after[1] @ kak.single_qubit_operations_before[1],
```

Comment:
Please add a clarifying comment explaining that matrix products are taken here because `_decomp_to_operations` implicitly inserts sqrt iswaps between pairs of single qubit unitaries.

Reply:
Done.